Adds Wireshark pipes support, cleanup OSX tcpdump handling.

micolous opened the pull request:
New:

* `WiresharkSink`, a PipeTools Sink for streaming packets to Wireshark
* `utils.get_temp_dir`: creates a temporary directory

Extensions:

* `tcpdump()`: Adds `use_tempfile`, to explicitly control use of a tempfile for packets.
* `tcpdump()`: Adds `read_stdin_opts` parameter, which allows callers to control the options used for reading packets over `stdin` (previously hard-coded to `-r -`).  This auto-detects `wireshark` to use `-ki -` instead.
* `tcpdump()`, `WrpcapSink`: Adds `linktype` parameter.
* `get_temp_file()`: Adds `fd` parameter, which allows a temporary file to be used without closing and re-opening it.
* Expands `wireshark()`, `Source` and `Pipe` documentation.
* `tdecode()`: Add `args` parameter (defaults to `-V`, as before), pass kwargs to `tcpdump()`.
* `QueueSink.recv`: add `block` and `timeout` parameters.

Fixes:

* `tcpdump()` now only uses a temporary file by default for calling tcpdump on OSX (to work around Apple's broken version of `tcpdump`). `stdin` is now used with other tools which are not impacted by this bug (eg: `tshark`).
* `wireshark()` now uses a pipe rather than a temporary file. Wireshark itself has options to save this file to disk if desired.
* `SniffSource` now only opens a socket on calling `start`.
* `WrpcapSink` only opens a `PcapWriter` on calling `start`.
* `RawPcapWriter` no longer writes a header on `write` calls with no packets specified.
* `RawPcapWriter` writes a header on `close` if no header has been written.
* `QueueSink.recv()` no longer busy-loops.
* `PcapWriter`, `wrpcap()`: support packets as bytes
* PEP-8 fixups and docstring revisions to touched methods.

Documentation fixes:

* Documentation wordsmithing for `wireshark()` and `Pipes`
* Use `py:function` / `py:class` for Sinks and `wireshark()` to allow easy cross-referencing, and expand this documentation.
* `wireshark()`: Remove references to `google.com`, and use RFC 5737 IP addresses instead
* `wireshark()`: Set a source IP address (to avoid looking up host's IP address).
* `wireshark()`: Remove `Ether` layer, as Wireshark works fine with `DLT_RAW` for `IP` packets, which removes the need to lookup MAC addresses
* Add `manpages_url` to configuration, to support `:manpage:` directives. This is pointing at Debian's server, which should have a good spread of versions of tools, and Debian generally adds manpages for tools that don't have them.

Test fixes:

* Removes `os.remove("test.png")` (which seems to be unused).
* Uses a temporary directory for some pipetool tests.

Other changes:

* `{Raw,}PcapWriter._write_packet`: Remove unused support for `packet` as tuple, as `write` will always unroll iterators for us (and do it better).
* `{Raw,}PcapWriter._write_packet`: Always set the `usec` parameter if `sec` was unset.
* `{Raw,}PcapWriter._write_packet`: Set `usec=0` if `sec` was set, but `usec` was unset (instead of using the current time's usec value

gpotter2 commented:
Really like this PR !

If you could rebase, it will fix the AppVeyor failure

gpotter2 commented on scapy/utils.py:
```diff
@@ -1365,12 +1390,9 @@ def import_hexcap():
-    wrpcap(f, pktlist, **kwargs)
-    with ContextManagerSubprocess("wireshark()", conf.prog.wireshark):
-        subprocess.Popen([conf.prog.wireshark, "-r", f])
+    tcpdump(pktlist, prog=conf.prog.wireshark)
```
Note that it is always nice to be able to send a `linklayer` argument in `wireshark`, when passing raw data for instance.. that’s mostly what the kwargs were for

Would be nice if you add support for the `-y` argument by default

micolous replied:
Ah, you mean `linktype`?

micolous replied:
OK, I've added this in 6634a22.

gpotter2 replied:
Yes sorry that's what I meant, thanks !

micolous commented:
Found another issue, tcpdump doesn't like integer linktypes...

micolous commented:
(in case it isn't clear already, please don't merge this yet...)

micolous commented:
I'm going to amend the docs a little more, squash and rebase this before it's ready.

micolous commented:
@gpotter2: I squashed fixups into commits that make a bit more sense.  I've updated the original change description with the extra changes since approval, please take a look!

gpotter2 commented:
the `winpcap=True` version already fails on some specific tests. It's using a very old libpcap version

About linktypes, we *could* hack our way through with something similar to
```python
>>> def hacky_linktype_name(value): 
...:     return next(k for k, v in six.iteritems(scapy.d...: ata.__dict__) if k[:3] == "DLT" and v == value)
>>>
>>> hacky_linktype_name(0)                              'DLT_NULL'
```

as a fallback, though that's a bit messy...

Otherwise, the doc improvement is great and PR looks good overall

gpotter2 commented on scapy/utils.py:
```diff
@@ -1253,35 +1278,48 @@ def _write_header(self, pkt):
-            for pkt in packet:
-                self._write_packet(pkt, sec=sec, usec=usec, caplen=caplen,
-                                   wirelen=wirelen)
-            return
```
I'll agree that tuple support is a weird thing we were doing. Though I'm not sure why it was there in the first place, that's a backward incompatible change.

I'll check it out to see how used it was
